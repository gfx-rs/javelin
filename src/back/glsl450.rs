use crate::{
    back::glsl_common::{write_type, Error, StatementBuilder},
    Binding, BuiltIn, FastHashMap, Module, TypeInner,
};
use std::{
    fmt::{self},
    io::Write,
};

pub fn write(module: &Module, out: &mut impl Write) -> Result<(), Error> {
    writeln!(out, "#version 450 core")?;

    let mut counter = 0;
    let mut names = FastHashMap::default();

    let mut namer = |name: Option<&String>| {
        if let Some(name) = name {
            names.insert(name.clone(), ());
            name.clone()
        } else {
            counter += 1;
            while names.get(&format!("_{}", counter)).is_some() {
                counter += 1;
            }
            format!("_{}", counter)
        }
    };

    let mut structs = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let name = namer(ty.name.as_ref());
                let mut idxs = Vec::new();

                for member in members {
                    idxs.push(namer(member.name.as_ref()))
                }

                structs.insert(handle, (name, idxs));
            }
            _ => continue,
        }
    }

    // Do a second pass to build the structs
    // TODO: glsl is order dependent so we need to build structs in order
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let (name, idxs) = structs.get(&handle).unwrap();

                writeln!(out, "struct {} {{", name)?;
                for (member, name) in members.iter().zip(idxs.iter()) {
                    writeln!(
                        out,
                        "   {} {}",
                        write_type(member.ty, &module.types, &structs)?,
                        name
                    )?;
                }
                writeln!(out, "}};")?;
            }
            _ => continue,
        }
    }

    let mut globals_lookup = FastHashMap::default();

    for (handle, global) in module.global_variables.iter() {
        if let Some(Binding::BuiltIn(built_in)) = global.binding {
            let semantic = match built_in {
                BuiltIn::Position => "gl_position",
                BuiltIn::GlobalInvocationId => "gl_GlobalInvocationID",
                BuiltIn::BaseInstance => "gl_BaseInstance",
                BuiltIn::BaseVertex => "gl_BaseVertex",
                BuiltIn::ClipDistance => "gl_ClipDistance",
                BuiltIn::InstanceIndex => "gl_InstanceIndex",
                BuiltIn::VertexIndex => "gl_VertexIndex",
                BuiltIn::PointSize => "gl_PointSize",
                BuiltIn::FragCoord => "gl_FragCoord",
                BuiltIn::FrontFacing => "gl_FrontFacing",
                BuiltIn::SampleIndex => "gl_SampleID",
                BuiltIn::FragDepth => "gl_FragDepth",
                BuiltIn::LocalInvocationId => "gl_LocalInvocationID",
                BuiltIn::LocalInvocationIndex => "gl_LocalInvocationIndex",
                BuiltIn::WorkGroupId => "gl_WorkGroupID",
            };

            globals_lookup.insert(handle, String::from(semantic));
            continue;
        }

        if let Some(ref binding) = global.binding {
            write!(out, "layout({}) ", binding.writer())?;
        }

        let name = namer(global.name.as_ref());

        writeln!(
            out,
            "{}{} {};",
            global.class.write_glsl(),
            write_type(global.ty, &module.types, &structs)?,
            name
        )?;

        globals_lookup.insert(handle, name);
    }

    let mut functions = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, func) in module.functions.iter() {
        functions.insert(handle, namer(func.name.as_ref()));
    }

    let mut types = module.types.clone();
    let mut typifier = crate::proc::Typifier::new();

    // TODO: glsl is order dependent so we need to build functions in order
    for (handle, func) in module.functions.iter() {
        let name = functions.get(&handle).unwrap();

        writeln!(
            out,
            "{} {}({}) {{",
            func.return_type
                .map_or(Ok(String::from("void")), |ty| write_type(
                    ty, &types, &structs
                ))?,
            name,
            func.parameter_types
                .iter()
                .map(|ty| write_type(*ty, &types, &structs))
                .collect::<Result<Vec<_>, _>>()?
                .join(","),
        )?;

        let mut builder = StatementBuilder {
            functions: &functions,
            globals: &globals_lookup,
            locals_lookup: &func
                .local_variables
                .iter()
                .map(|(handle, local)| (handle, namer(local.name.as_ref())))
                .collect(),
            structs: &structs,
            args: &func
                .parameter_types
                .iter()
                .enumerate()
                .map(|(pos, _)| (pos as u32, namer(None)))
                .collect(),
            expressions: &func.expressions,
            types: &mut types,
            locals: &func.local_variables,
            typifier: &mut typifier,
        };

        for sta in func.body.iter() {
            writeln!(out, "{}", sta.write_glsl(module, &mut builder)?)?;
        }

        writeln!(out, "}}")?;
    }

    Ok(())
}

struct BindingWriter<'a> {
    inner: &'a Binding,
}

impl Binding {
    fn writer<'a>(&'a self) -> BindingWriter<'a> {
        BindingWriter { inner: self }
    }
}

impl<'a> fmt::Display for BindingWriter<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.inner {
            Binding::BuiltIn(_) => write!(f, ""), // Ignore because they are variables with a predefined name
            Binding::Location(location) => write!(f, "location={}", location),
            Binding::Descriptor { set, binding } => write!(f, "set={},binding={}", set, binding),
        }
    }
}
