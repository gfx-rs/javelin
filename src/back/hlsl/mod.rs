mod keywords;
mod writer;

use std::fmt::Error as FmtError;
use thiserror::Error;

pub use writer::Writer;

#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
pub struct ShaderModel(u16);

impl Default for ShaderModel {
    fn default() -> Self {
        ShaderModel(50)
    }
}

#[derive(Error, Debug)]
pub enum Error {
    #[error(transparent)]
    IoError(#[from] FmtError),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedShaderModel(ShaderModel),
    #[error("A scalar with an unsupported width was requested: {0:?} {1:?}")]
    UnsupportedScalar(crate::ScalarKind, crate::Bytes),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedBuiltIn(crate::BuiltIn),
    #[error("{0}")]
    Unimplemented(String), // TODO: Error used only during development
}

pub fn write_string(
    module: &crate::Module,
    info: &crate::valid::ModuleInfo,
    shader_model: ShaderModel,
) -> Result<String, Error> {
    let mut w = Writer::new(String::new(), shader_model);
    w.write(module, info)?;
    let output = w.finish();
    Ok(output)
}
