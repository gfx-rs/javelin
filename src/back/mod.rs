//! Functions which export shader modules into binary and text formats.

#[cfg(feature = "glsl-out")]
pub mod glsl;
#[cfg(feature = "msl-out")]
pub mod msl;
#[cfg(feature = "spv-out")]
pub mod spv;

impl crate::Expression {
    /// Returns the ref count, upon reaching which this expression
    /// should be considered for baking.
    #[allow(dead_code)]
    fn bake_ref_count(&self) -> usize {
        match *self {
            // accesses are never cached, only loads are
            crate::Expression::Access { .. } | crate::Expression::AccessIndex { .. } => !0,
            // image operations look better when isolated
            crate::Expression::ImageSample { .. } | crate::Expression::ImageLoad { .. } => 1,
            // cache expressions that are referenced multiple times
            _ => 2,
        }
    }
}
