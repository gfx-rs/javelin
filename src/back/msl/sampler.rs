#[derive(Clone, Debug, PartialEq)]
pub enum Coord {
    Normalized,
    Pixel,
}

impl Default for Coord {
    fn default() -> Self {
        Self::Normalized
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Address {
    Repeat,
    MirroredRepeat,
    ClampToEdge,
    ClampToZero,
    ClampToBorder,
}

impl Default for Address {
    fn default() -> Self {
        Self::ClampToEdge
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum BorderColor {
    TransparentBlack,
    OpaqueBlack,
    OpaqueWhite,
}

impl Default for BorderColor {
    fn default() -> Self {
        Self::TransparentBlack
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Filter {
    Nearest,
    Linear,
}

impl Default for Filter {
    fn default() -> Self {
        Self::Nearest
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum CompareFunc {
    Never,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    Equal,
    NotEqual,
    Always,
}

impl Default for CompareFunc {
    fn default() -> Self {
        Self::Never
    }
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct InlineSampler {
    pub coord: Coord,
    pub address: [Address; 3],
    pub border_color: BorderColor,
    pub mag_filter: Filter,
    pub min_filter: Filter,
    pub mip_filter: Option<Filter>,
    pub compare_func: CompareFunc,
}
