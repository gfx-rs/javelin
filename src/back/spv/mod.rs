mod layout;
pub mod parser;

use spirv::*;

struct Module {
    physical_layout: PhysicalLayout,
    logical_layout: LogicalLayout,
}

impl Module {
    fn new(header: &crate::Header) -> Self {
        Module {
            physical_layout: PhysicalLayout::new(header, 0x0u32),
            logical_layout: LogicalLayout::new(),
        }
    }
}

struct PhysicalLayout {
    magic_number: Word,
    version: Word,
    generator: Word,
    bound: Word,
    instruction_schema: Word,
}

impl PhysicalLayout {
    fn new(header: &crate::Header, instruction_schema: Word) -> Self {
        let version: Word = (0x0u32 << 24)
            | ((header.version.0 as u32) << 16)
            | ((header.version.1 as u32) << 8)
            | header.version.2 as u32;

        PhysicalLayout {
            magic_number: MAGIC_NUMBER,
            version,
            generator: header.generator,
            bound: 0,
            instruction_schema,
        }
    }

    fn in_words(&self) -> Vec<Word> {
        let mut words: Vec<Word> = vec![];

        words.push(self.magic_number);
        words.push(self.version);
        words.push(self.generator);

        match self.bound {
            0 => panic!("Bound cannot be 0"),
            _ => words.push(self.bound),
        }

        words.push(self.instruction_schema);
        words
    }
}

pub(crate) struct LogicalLayout {
    capabilities: Vec<Word>,
    extensions: Vec<Word>,
    ext_inst_imports: Vec<Word>,
    memory_model: Vec<Word>,
    entry_points: Vec<Word>,
    execution_modes: Vec<Word>,
    debugs: Vec<Word>,
    annotations: Vec<Word>,
    type_declarations: Vec<Word>,
    constants: Vec<Word>,
    global_variables: Vec<Word>,
    function_declarations: Vec<Word>,
    function_definitions: Vec<Word>,
}

impl LogicalLayout {
    fn new() -> Self {
        LogicalLayout {
            capabilities: vec![],
            extensions: vec![],
            ext_inst_imports: vec![],
            memory_model: vec![],
            entry_points: vec![],
            execution_modes: vec![],
            debugs: vec![],
            annotations: vec![],
            type_declarations: vec![],
            constants: vec![],
            global_variables: vec![],
            function_declarations: vec![],
            function_definitions: vec![],
        }
    }

    fn in_words(&self) -> Vec<u32> {
        let mut words: Vec<Word> = vec![];

        for capability in self.capabilities.iter() {
            words.push(*capability);
        }

        for extension in self.extensions.iter() {
            words.push(*extension);
        }

        for ext_inst_import in self.ext_inst_imports.iter() {
            words.push(*ext_inst_import);
        }

        for memory_model in self.memory_model.iter() {
            words.push(*memory_model);
        }

        for entry_point in self.entry_points.iter() {
            words.push(*entry_point);
        }

        for execution_mode in self.execution_modes.iter() {
            words.push(*execution_mode);
        }

        for debug in self.debugs.iter() {
            words.push(*debug);
        }

        for annotation in self.annotations.iter() {
            words.push(*annotation);
        }

        for type_declaration in self.type_declarations.iter() {
            words.push(*type_declaration);
        }

        for constants in self.constants.iter() {
            words.push(*constants);
        }

        for global_variables in self.global_variables.iter() {
            words.push(*global_variables);
        }

        for function_declaration in self.function_declarations.iter() {
            words.push(*function_declaration);
        }

        for function_definition in self.function_definitions.iter() {
            words.push(*function_definition);
        }

        words
    }
}

struct Instruction {
    op: Op,
    wc: u32,
    type_id: Option<Word>,
    result_id: Option<Word>,
    operands: Vec<Word>,
}

impl Instruction {
    fn new(op: Op) -> Self {
        Instruction {
            op,
            wc: 1, // Always start at 1 for the first word (OP + WC),
            type_id: None,
            result_id: None,
            operands: vec![],
        }
    }

    fn set_type(&mut self, id: Word) {
        assert!(self.type_id.is_none(), "Type can only be set once");
        self.type_id = Some(id);
        self.wc += 1;
    }

    fn set_result(&mut self, id: Word) {
        assert!(self.result_id.is_none(), "Result can only be set once");
        self.result_id = Some(id);
        self.wc += 1;
    }

    fn add_operand(&mut self, operand: Word) {
        self.operands.push(operand);
        self.wc += 1;
    }

    fn add_operands(&mut self, operands: Vec<Word>) {
        for operand in operands {
            self.add_operand(operand);
        }
    }

    fn to_words(&self) -> Vec<Word> {
        let mut words = Vec::with_capacity(self.wc as usize);

        let wc_op = (self.wc << 16 | self.op as u32) as u32;

        words.push(wc_op);

        if self.type_id.is_some() {
            words.push(self.type_id.unwrap());
        }

        if self.result_id.is_some() {
            words.push(self.result_id.unwrap());
        }

        for operand in self.operands.iter() {
            words.push(*operand);
        }

        words
    }
}
