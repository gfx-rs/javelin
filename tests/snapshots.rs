use serde::Deserialize;

#[derive(Hash, PartialEq, Eq, Deserialize)]
enum Stage {
    Vertex,
    Fragment,
    Compute,
}

#[derive(Hash, PartialEq, Eq, Deserialize)]
struct BindSource {
    stage: Stage,
    group: u32,
    binding: u32,
}

#[derive(Deserialize)]
struct BindTarget {
    #[serde(default)]
    buffer: Option<u8>,
    #[serde(default)]
    texture: Option<u8>,
    #[serde(default)]
    sampler: Option<u8>,
    #[serde(default)]
    mutable: bool,
}

#[derive(Default, Deserialize)]
struct Parameters {
    #[serde(default)]
    spv_flow_dump_prefix: String,
    #[cfg_attr(not(feature = "spv-out"), allow(dead_code))]
    spv_capabilities: naga::FastHashSet<spirv::Capability>,
    #[cfg_attr(not(feature = "msl-out"), allow(dead_code))]
    mtl_bindings: naga::FastHashMap<BindSource, BindTarget>,
}

#[cfg(feature = "msl-out")]
fn check_output_msl(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::msl;
    let mut binding_map = msl::BindingMap::default();
    for (key, value) in params.mtl_bindings.iter() {
        binding_map.insert(
            msl::BindSource {
                stage: match key.stage {
                    Stage::Vertex => naga::ShaderStage::Vertex,
                    Stage::Fragment => naga::ShaderStage::Fragment,
                    Stage::Compute => naga::ShaderStage::Compute,
                },
                group: key.group,
                binding: key.binding,
            },
            msl::BindTarget {
                buffer: value.buffer,
                texture: value.texture,
                sampler: value.sampler,
                mutable: value.mutable,
            },
        );
    }
    let options = msl::Options {
        lang_version: (1, 0),
        spirv_cross_compatibility: false,
        binding_map,
    };
    let (msl, _) = msl::write_string(&module, &options).unwrap();
    insta::assert_snapshot!(format!("{}{}", name, ".msl"), msl);
}

#[cfg(feature = "spv-out")]
fn check_output_spv(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::spv;
    use rspirv::binary::Disassemble;
    let spv = spv::write_vec(
        &module,
        spv::WriterFlags::NONE,
        params.spv_capabilities.clone(),
    )
    .unwrap();
    let dis = rspirv::dr::load_words(spv)
        .expect("Produced invalid SPIR-V")
        .disassemble();
    insta::assert_snapshot!(format!("{}{}", name, ".spvasm"), dis);
}

#[cfg(feature = "wgsl-in")]
fn convert_wgsl(name: &str) {
    let params =
        match std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".param.ron")) {
            Ok(string) => ron::de::from_str(&string).expect("Couldn't find param file"),
            Err(_) => Parameters::default(),
        };

    let module = naga::front::wgsl::parse_str(
        &std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".wgsl"))
            .expect("Couldn't find wgsl file"),
    )
    .unwrap();
    naga::proc::Validator::new().validate(&module).unwrap();

    #[cfg(feature = "msl-out")]
    check_output_msl(&module, name, &params);
    #[cfg(feature = "spv-out")]
    check_output_spv(&module, name, &params);
}

#[cfg(feature = "wgsl-in")]
#[test]
fn converts_wgsl_quad() {
    convert_wgsl("quad");
}
