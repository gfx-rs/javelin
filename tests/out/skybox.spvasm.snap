---
source: tests/snapshots.rs
expression: dis
---
; SPIR-V
; Version: 1.0
; Generator: rspirv
; Bound: 116
OpCapability Shader
%1 = OpExtInstImport "GLSL.std.450"
OpMemoryModel Logical GLSL450
OpEntryPoint Vertex %37 "vs_main" %30 %33 %35
OpEntryPoint Fragment %108 "fs_main" %101 %104 %107
OpExecutionMode %108 OriginUpperLeft
OpSource GLSL 450
OpName %11 "Data"
OpMemberName %11 0 "proj_inv"
OpMemberName %11 1 "view"
OpName %10 "r_data"
OpName %15 "r_texture"
OpName %18 "r_sampler"
OpName %21 "tmp1"
OpName %23 "tmp2"
OpName %24 "out"
OpName %25 "VertexOutput"
OpName %30 "vertex_index"
OpName %33 "position"
OpName %35 "uv"
OpName %37 "vs_main"
OpName %37 "vs_main"
OpName %101 "position"
OpName %104 "uv"
OpName %108 "fs_main"
OpName %108 "fs_main"
OpDecorate %11 Block
OpMemberDecorate %11 0 Offset 0
OpMemberDecorate %11 0 ColMajor
OpMemberDecorate %11 0 MatrixStride 16
OpMemberDecorate %11 1 Offset 64
OpMemberDecorate %11 1 ColMajor
OpMemberDecorate %11 1 MatrixStride 16
OpDecorate %10 DescriptorSet 0
OpDecorate %10 Binding 0
OpDecorate %15 DescriptorSet 0
OpDecorate %15 Binding 1
OpDecorate %18 DescriptorSet 0
OpDecorate %18 Binding 2
OpDecorate %30 BuiltIn VertexIndex
OpDecorate %33 BuiltIn Position
OpDecorate %35 Location 0
OpDecorate %101 BuiltIn FragCoord
OpDecorate %104 Location 0
OpDecorate %107 Location 0
%2 = OpTypeVoid
%4 = OpTypeInt 32 1
%3 = OpConstant  %4  2
%5 = OpConstant  %4  1
%7 = OpTypeFloat 32
%6 = OpConstant  %7  4.0
%8 = OpConstant  %7  1.0
%9 = OpConstant  %7  0.0
%13 = OpTypeVector %7 4
%12 = OpTypeMatrix %13 4
%11 = OpTypeStruct %12 %12
%14 = OpTypePointer Uniform %11
%10 = OpVariable  %14  Uniform
%16 = OpTypeImage %7 Cube 0 0 0 1 Unknown
%17 = OpTypePointer UniformConstant %16
%15 = OpVariable  %17  UniformConstant
%19 = OpTypeSampler
%20 = OpTypePointer UniformConstant %19
%18 = OpVariable  %20  UniformConstant
%22 = OpTypePointer Function %4
%26 = OpTypeVector %7 3
%25 = OpTypeStruct %13 %26
%27 = OpTypePointer Function %25
%29 = OpTypeInt 32 0
%31 = OpTypePointer Input %29
%30 = OpVariable  %31  Input
%34 = OpTypePointer Output %13
%33 = OpVariable  %34  Output
%36 = OpTypePointer Output %26
%35 = OpVariable  %36  Output
%38 = OpTypeFunction %2
%53 = OpTypePointer Uniform %12
%54 = OpConstant  %4  1
%62 = OpConstant  %4  1
%70 = OpConstant  %4  1
%78 = OpTypeMatrix %26 3
%81 = OpConstant  %4  0
%85 = OpTypePointer Function %26
%91 = OpConstant  %4  1
%93 = OpTypePointer Function %13
%94 = OpConstant  %4  0
%102 = OpTypePointer Input %13
%101 = OpVariable  %102  Input
%105 = OpTypePointer Input %26
%104 = OpVariable  %105  Input
%107 = OpVariable  %34  Output
%113 = OpTypeSampledImage %16
%37 = OpFunction  %2  None %38
%28 = OpLabel
%21 = OpVariable  %22  Function
%23 = OpVariable  %22  Function
%24 = OpVariable  %27  Function
%32 = OpLoad  %29  %30
OpBranch %39
%39 = OpLabel
%40 = OpBitcast  %4  %32
%41 = OpSDiv  %4  %40 %3
OpStore %21 %41
%42 = OpBitcast  %4  %32
%43 = OpBitwiseAnd  %4  %42 %5
OpStore %23 %43
%44 = OpLoad  %4  %21
%45 = OpConvertSToF  %7  %44
%46 = OpFMul  %7  %45 %6
%47 = OpFSub  %7  %46 %8
%48 = OpLoad  %4  %23
%49 = OpConvertSToF  %7  %48
%50 = OpFMul  %7  %49 %6
%51 = OpFSub  %7  %50 %8
%52 = OpCompositeConstruct  %13  %47 %51 %9 %8
%55 = OpAccessChain  %53  %10 %54
%56 = OpLoad  %12  %55
%57 = OpCompositeExtract  %13  %56 0
%58 = OpCompositeExtract  %7  %57 0
%59 = OpCompositeExtract  %7  %57 1
%60 = OpCompositeExtract  %7  %57 2
%61 = OpCompositeConstruct  %26  %58 %59 %60
%63 = OpAccessChain  %53  %10 %62
%64 = OpLoad  %12  %63
%65 = OpCompositeExtract  %13  %64 1
%66 = OpCompositeExtract  %7  %65 0
%67 = OpCompositeExtract  %7  %65 1
%68 = OpCompositeExtract  %7  %65 2
%69 = OpCompositeConstruct  %26  %66 %67 %68
%71 = OpAccessChain  %53  %10 %70
%72 = OpLoad  %12  %71
%73 = OpCompositeExtract  %13  %72 2
%74 = OpCompositeExtract  %7  %73 0
%75 = OpCompositeExtract  %7  %73 1
%76 = OpCompositeExtract  %7  %73 2
%77 = OpCompositeConstruct  %26  %74 %75 %76
%79 = OpCompositeConstruct  %78  %61 %69 %77
%80 = OpTranspose  %78  %79
%82 = OpAccessChain  %53  %10 %81
%83 = OpLoad  %12  %82
%84 = OpMatrixTimesVector  %13  %83 %52
%86 = OpCompositeExtract  %7  %84 0
%87 = OpCompositeExtract  %7  %84 1
%88 = OpCompositeExtract  %7  %84 2
%89 = OpCompositeConstruct  %26  %86 %87 %88
%90 = OpMatrixTimesVector  %26  %80 %89
%92 = OpAccessChain  %85  %24 %91
OpStore %92 %90
%95 = OpAccessChain  %93  %24 %94
OpStore %95 %52
%96 = OpLoad  %25  %24
%97 = OpCompositeExtract  %13  %96 0
OpStore %33 %97
%98 = OpCompositeExtract  %26  %96 1
OpStore %35 %98
OpReturn
OpFunctionEnd
%108 = OpFunction  %2  None %38
%99 = OpLabel
%103 = OpLoad  %13  %101
%106 = OpLoad  %26  %104
%100 = OpCompositeConstruct  %25  %103 %106
%109 = OpLoad  %16  %15
%110 = OpLoad  %19  %18
OpBranch %111
%111 = OpLabel
%112 = OpCompositeExtract  %26  %100 1
%114 = OpSampledImage  %113  %109 %110
%115 = OpImageSampleImplicitLod  %13  %114 %112
OpStore %107 %115
OpReturn
OpFunctionEnd
